datafusion/sql/src/values.rs

```diff
@@ -41,6 +41,11 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                     .collect::<Result<Vec<_>>>()
             })
             .collect::<Result<Vec<_>>>()?;
-        LogicalPlanBuilder::values(values)?.build()
+        let column_types = planner_context.values_column_data_types();
+        if column_types.is_empty() {
+            LogicalPlanBuilder::values(values)?.build()
+        } else {
+            LogicalPlanBuilder::values_with_types(values, column_types)?.build()
```

Comment:
I am not convinced this is a good idea, as it will lead to inconsistent behavior or statements that should be equivalent. See https://github.com/apache/datafusion/issues/12103#issuecomment-2307760145

Reply:
I agree with @findepi  -- https://github.com/apache/datafusion/issues/12103#issuecomment-2312601583 

Thank you @davisp  for this PR -- let us know what you think about the alternate proposal

Reply:
I'm definitely on board in making this more general. It didn't feel super awesome to be fixing things this specifically in the first place, it was just the first thing that came to mind.

@alamb You mentioned in your comment on #12103 that there are other coercion patterns to follow, can you point me at anything vaguely similar to reference? Skimming the DataType docs, I'm not immediately seeing how I'd handle the expansion correctly.

Reply:
In general, the coercsion code is in https://github.com/apache/datafusion/tree/main/datafusion/expr/src/type_coercion

This has a reasonable description:
https://docs.rs/datafusion/latest/datafusion/logical_expr/type_coercion/index.html